doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +748,459 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
 
-Notations
-~~~~~~~~~
+   .. insertprodn ltac2_branches atomic_tac2pat
 
-The Ltac2 parser can be extended with syntactic notations.
+   .. prodn::
+      ltac2_branches ::= {? %| } {+| @tac2pat1 => @ltac2_expr }
+      tac2pat1 ::= @qualid {+ @tac2pat0 }
+      | @qualid
+      | [ ]
+      | @tac2pat0 :: @tac2pat0
+      | @tac2pat0
+      tac2pat0 ::= _
+      | ()
+      | @qualid
+      | ( {? @atomic_tac2pat } )
+      atomic_tac2pat ::= @tac2pat1 : @ltac2_type
+      | @tac2pat1 , {*, @tac2pat1 }
+      | @tac2pat1
 
-.. cmd:: Ltac2 Notation {+ {| @lident (@ltac2_scope) | @string } } {? : @int} := @ltac2_term
+.. note::
+
+   For now, deep pattern matching is not implemented.
+
+
+.. _ltac2_notations:
+
+Notations
+---------
+
+.. cmd:: Ltac2 Notation @string {* @name ( @ltac2_scope ) } {? : @int } := @ltac2_expr
    :name: Ltac2 Notation
 
-   A Ltac2 notation adds a parsing rule to the Ltac2 grammar, which is expanded
+   .. todo seems like name maybe should use lident rather than ident, considering:
+
+      Ltac2 Notation "ex1" X(constr) := print (of_constr X).
+      ex1 1.
+
+      Unbound constructor X
+
+      This works fine with lower-case "x" in place of "X"
+
+   .. todo Ltac2 Notation := permits redefining same symbol (no warning)
+      Also allows defining a symbol beginning with uppercase, which is prohibited
+      in similar constructs.
+
+   :cmd:`Ltac2 Notation` provides a way to extend |Coq|'s syntax.  The left-hand
+   side (before the `:=`) defines the syntax to recognize and gives formal parameter
+   names for the syntactic values.  When the notation is used, the values are substituted
+   into the right-hand side.  In the following example, `x` is the formal parameter name, `constr` is
+   the name of a nonterminal describing the syntax (which the documentation calls :token:`term`).
+   `print` and `of_constr` are functions provided by |Coq| through `Message.v`.
+
+   .. todo "print" doesn't seem to pay attention to "Set Printing All"
+
+   .. example:: Printing a :n:`@term`
+
+      .. coqtop:: none
+
+         Goal True.
+
+      .. coqtop:: all
+
+         From Ltac2 Require Import Message.
+         Ltac2 Notation "ex1" x(constr) := print (of_constr x).
+         ex1 (1 + 2).
+
+   There are also metasyntactic classes described :ref:`here<syntactic_classes>`
+   that combine other items.  For example, `list1(constr, ",")`
+   recognizes a comma-separated list of one or more :token:`term`\s.
+
+   .. example:: Parsing a list of :n:`@term`\s
+
+      .. coqtop:: abort all
+
+         Ltac2 rec print_list x := match x with
+         | a :: t => print (of_constr a); print_list t
+         | [] => ()
+         end.
+         Ltac2 Notation "ex2" x(list1(constr, ",")) := print_list x.
+         ex2 1, 2, 3.
+
+   An Ltac2 notation adds a parsing rule to the Ltac2 grammar, which is expanded
    to the provided body where every token from the notation is let-bound to the
    corresponding generated expression.
```

Comment:
Maybe add a note that, much like in Gallina, the RHS is *not* typechecked until substitution time?